### DISCUSSION TITLE
SSL support / private keys

### DISCUSSION THREAD

**Discussion by ptitzler** [elyra-ai/elyra#2828] [Internal]
To support #2787, #2812, and #2797
  - Elyra needs to provide access to certificate locations whenever an outbound HTTP request is made
  - The user/administrator needs to configure the locations and make the certificates available

This could be accomplished as follows:
 -  augment all `requests.XXX` requests to optionally include the [location of a certificate file or directory](https://requests.readthedocs.io/en/latest/user/advanced/#ssl-cert-verification):
     - HOW
        - explicitly -> `requests.get('https://domain.com', verify='/path/to/certfile')`
        -  implicitly -> set `REQUESTS_CA_BUNDLE` environment variable
     - WHERE
        - JupyterLab
        - generic component
 - 'register' certificate file location
     - HOW
        - re-use  `REQUESTS_CA_BUNDLE` environment variable (aka piggy-back on requests' solution)
        - re-use [JupyterLab configuration setting](https://jupyter-server.readthedocs.io/en/latest/operators/public-server.html?#securing-a-jupyter-server) (h/t @kevin-bates )
        - introduce custom environment variable
     - WHERE
        - JupyterLab (might require optional volume mount that exposes the certificates)
        - generic component  (might require optional configmap, secret or volume mount that exposes the certificates)
        - custom components  (might require optional configmap, secret or volume mount  that exposes the certificates)

**Comment by ptitzler** [comment]
@kevin-bates  re: 're-use [JupyterLab configuration setting](https://jupyter-server.readthedocs.io/en/latest/operators/public-server.html?#securing-a-jupyter-server)' (e.g `c.ServerApp.certfile`) this wouldn't be available to our backend code if one were to use the CLI to (for example) run a pipeline since no running JL instance is required, right?

**Reply by kevin-bates** [reply]
That is correct.  I suppose we could look into loading that configuration to access the value of `certfile` or introduce a CLI option (since our CLI tools don't really leverage the Jupyter config framework (i.e., traitlets)) that is documented to be the same value as `c.ServerApp.certfile`.
